core/classifiers/CodeClassifier/CodeClassifier.py

```diff
@@ -0,0 +1,710 @@
+
+    # Add 'UNK' in token_to_id. This will be used to replace any token
+    # occurring in program which is not in valid_token.
+    token_to_id[UNK_TOKEN] = len(token_to_id)
+
+    # Add 'V' in token_to_id. This token will be used to replace all
+    # variables and methods in program.
+    token_to_id[VAR_TOKEN] = len(token_to_id)
+    return token_to_id
+
+
+def tokenize_data(training_data, threshold=VOCABULARY_THRESHOLD):
+    """Tokenize Python programs in dataset for winnowing.
+    Args:
+        training_data: dict. A dictionary containing training_data. Structure of
+            training_data is as follows,
+                {
+                    ID: {
+                        'source': str. Source code of program.
+                        'class': int. Associated feedback class with program
+                    } for each program in training_data.
+                }
+        threshold: int. Ignore a token if it appears less than 'threshold' times
+            in dataset.
+
+    Returns:
+        dict. A dictionary containing training data with additional key 'tokens'
+        for each data point which stores tokens of corresponding data point.
+        Structure of dict is as follows,
+            {
+                ID: {
+                    'source': str. Source code of program.
+                    'class': int. Associated feedback class with program.
+                    'tokens': list(str). Token list of program.
+                } for each program in training_data.
+            }
+    """
+    token_to_id = map_tokens_to_ids(training_data, threshold)
+
+    # Tokenize all programs in dataset.
+    for program_id in training_data:
+        program = training_data[program_id]['source']
+        token_program = []
+        for token_id, token_name in get_tokens(program):
+            # Ignore all newline, comment and empty string tokens.
+            if (token_id == token.N_TOKENS or token_id == COMMENT_TOKEN
+                    or token_name.strip() == ''):
+                continue
+            elif token_id == token.NAME:
+                if token_name in keyword.kwlist:
+                    # If token_id is tokens.NAME then only add if it is a python
+                    # keyword and is present in token_to_id else replace it
+                    # with UNK_TOKEN.
+                    if token_name in token_to_id:
+                        token_program.append(token_name)
+                    else:
+                        token_program.append(UNK_TOKEN)
+                else:
+                    # If token_id is tokens.NAME and it is not a python keyword
+                    # then it is a variable or method.
+                    # Treat all methods and variables same.
+                    token_program.append(VAR_TOKEN)
+            else:
+                # Add token only if it present in token_to_id. Otherwise replace
+                # the token with UNK_TOKEN.
+                if token_name in token_to_id:
```

Comment:
This seems to match the logic in elif/if above. Maybe combine? So the logic is:

    if ignorable:
        continue
    elif token_id == token.NAME and token_name not in keyword.kwlist:
        ...
    else:
        ...